### ISSUE TITLE
Cluster Sharding Serialiser occasionally failing when using Hyperion

### ISSUE THREAD

**Issue by andyfurnival** [akkadotnet/akka.net#5116]
**Version Information**
Version of Akka.NET? 
1.4.21
Which Akka.NET Modules? 
Akka.Cluster.Sharding
Akka.Persistence
Hyperion

**Describe the bug**
Occasionally, I'm getting exceptions on Coordinator and Shard replaying events (using Redis Persistence store), with the error message "System.Runtime.Serialization.SerializationException: Offset must be specified in whole minutes. (Parameter 'offset')" or "System.IndexOutOfRangeException: Index was outside the bounds of the array.". This is being thrown when the coordinator / shard is using Hyperion serialiser. I don't specify which serialiser to use for Cluster Sharding, its taken from the Default Config in Akka.Cluster.Sharding which looks to use Protobuf, but I do specify in the last fallback of my Akka configuration, my Journal and Snapshot serialisation which are all Protobuf

Full stack:
```
System.Runtime.Serialization.SerializationException: Offset must be specified in whole minutes. (Parameter 'offset')
 ---> System.ArgumentException: Offset must be specified in whole minutes. (Parameter 'offset')
   at System.DateTimeOffset.ValidateOffset(TimeSpan offset)
   at Hyperion.ValueSerializers.DateTimeOffsetSerializer.ReadDateTimeOffset(Stream stream, Byte[] bytes)
   at Hyperion.ValueSerializers.SessionAwareByteArrayRequiringValueSerializer`1.ReadValue(Stream stream, DeserializerSession session)
   at Hyperion.Serializer.Deserialize[T](Stream stream)
   at Akka.Serialization.HyperionSerializer.FromBinary(Byte[] bytes, Type type)
   --- End of inner exception stack trace ---
   at Akka.Serialization.HyperionSerializer.FromBinary(Byte[] bytes, Type type)
   at Akka.Serialization.Serializer.FromBinary[T](Byte[] bytes)
   at System.Linq.Enumerable.SelectArrayIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.OrderedEnumerable`1.TryGetFirst(Boolean& found)
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at Akka.Persistence.Redis.Snapshot.RedisSnapshotStore.LoadAsync(String persistenceId, SnapshotSelectionCriteria criteria)
   at Akka.Util.Internal.AtomicState.CallThrough[T](Func`1 task)
   at Akka.Util.Internal.AtomicState.CallThrough[T](Func`1 task) 

Persistence failure when replaying events for persistenceId ["/system/sharding/EventCoordinator/singleton/coordinator"]. Last known sequence number [0]
```
And this message
```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at lambda_method173(Closure , Stream , DeserializerSession )
   at Hyperion.Serializer.Deserialize[T](Stream stream)
   at Akka.Serialization.HyperionSerializer.FromBinary(Byte[] bytes, Type type)
   at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest)
   at Akka.Persistence.Serialization.PersistenceMessageSerializer.GetPayload(PersistentPayload payload)
   at Akka.Persistence.Serialization.PersistenceMessageSerializer.GetPersistentRepresentation(PersistentMessage message)
   at Akka.Serialization.Serialization.<>c__DisplayClass23_0.<Deserialize>b__0()
   at Akka.Serialization.Serialization.WithTransport[T](Func`1 action)
   at Akka.Persistence.Redis.JournalHelper.PersistentFromBytes(Byte[] bytes)
   at Akka.Persistence.Redis.Journal.RedisJournal.ReplayMessagesAsync(IActorContext context, String persistenceId, Int64 fromSequenceNr, Int64 toSequenceNr, Int64 max, Action`1 recoveryCallback) 

Persistence failure when replaying events for persistenceId ["Event-100006"]. Last known sequence number [2] 

```
**To Reproduce**
As this is happening in Akka internals, I'm not sure how I can reproduce this, even running on our infrastructure, its not consistently happening.  Sometimes I see I had a circuit breaker trigger on Redis (as recovery was causing Redis to struggle).

```
                var config = ConfigurationFactory.ParseString(File.ReadAllText("HOCON"))  #main HOCON 
                    .BootstrapFromDocker()
                    .WithFallback(DistributedPubSub.DefaultConfig())
                    .WithFallback(ClusterSharding.DefaultConfig()). #Akka.Net defaults to Protobuf
                    .WithFallback(SnapshotSerializer.Config)           #My Protobuf serialisation in separate HOCON
                    .WithFallback(ProtoMessageSerializer.Config);  #My Protobuf serialisation in separate HOCON
```

My main HOCON for serialisation

```
serializers {
            akka-pubsub = "Akka.Cluster.Tools.PublishSubscribe.Serialization.DistributedPubSubMessageSerializer, Akka.Cluster.Tools"
            hyperion = "Akka.Serialization.HyperionSerializer, Akka.Serialization.Hyperion"
            akka-cluster-client = "Akka.Cluster.Tools.Client.Serialization.ClusterClientMessageSerializer, Akka.Cluster.Tools"
            phobos-tracing = "Phobos.Tracing.Serialization.TraceEnvelopeSerializer, Phobos.Tracing"
          }
        serialization-bindings {
            "Akka.Cluster.Tools.PublishSubscribe.IDistributedPubSubMessage, Akka.Cluster.Tools" = akka-pubsub
            "Akka.Cluster.Tools.PublishSubscribe.Internal.SendToOneSubscriber, Akka.Cluster.Tools" = akka-pubsub
            "Akka.Cluster.Tools.Client.IClusterClientMessage, Akka.Cluster.Tools" = akka-cluster-client
            "Phobos.Tracing.IWithTrace, Phobos.Tracing" = phobos-tracing
            "System.Object" = hyperion
          }
        serialization-identifiers {
            "Akka.Cluster.Tools.PublishSubscribe.Serialization.DistributedPubSubMessageSerializer, Akka.Cluster.Tools" = 9            
            "Akka.Cluster.Tools.Client.Serialization.ClusterClientMessageSerializer, Akka.Cluster.Tools" = 15
            "Phobos.Tracing.Serialization.TraceEnvelopeSerializer, Phobos.Tracing" = 210
          }
```
**Expected behavior**
Coordinator / Shard recovery deserialisation should be performed using Protobuf

**Actual behavior**
Coordinator / Shard recovery deserialisation occasionally performed using Hyperion

**Screenshots**
See stack trace

**Environment**
Dotnet 5 / Docker

**Additional context**
I do have system.object using Hyperion as a fallback in HOCON, I mostly use protobuf for all serialisation using persistence and internal messages, but some do rely on system defaults as we migrate away from Hyperion completely - not sure if having Hyperion here causes the system to fall back to Hyperion on certain conditions

**Comment by ismaelhamed** [comment]
Yes, `"System.Object" = hyperion` is a catch-all fallback, so all *objects that do not have a specified serializer* will be serialized now using Hyperion instead of the default Newtonsoft.Json.

Have you specified a custom serializer for messages persisted via Redis? I can't see it in the config. Because AFAIK the Redis plugin will use the configured serializer (Hyperion in this case) for the payload of the messages.

**Comment by Aaronontheweb** [comment]
Looks like this issue is coming from the deserialized Akka.Cluster.Sharding payloads being written into Akka.Persistence.Redis - those _should_ have their own dedicated protobuf serializer, if I am not mistaken.

```
Persistence failure when replaying events for persistenceId ["/system/sharding/EventCoordinator/singleton/coordinator"]. Last known sequence number [0]
```

@Arkatufus could you look into this and see if you can reproduce this? Specifically, are those sharding snapshots being saved using the `object` serializer rather than something custom inside Akka.Cluster.Sharding when running sharding in persistence mode?

**Comment by Arkatufus** [comment]
@andyfurnival can you post your complete HOCON configuration?

**Comment by Arkatufus** [comment]
@andyfurnival To be clear, the error was raised because you're expecting the message to be deserialized using your custom protobuf deserializer and not Hyperion?

**Comment by andyfurnival** [comment]
> To be clear, the error was raised because you're expecting the message to be deserialized using your custom protobuf deserializer and not Hyperion
@Arkatufus no, I'm expecting that Akka.net's protobuf serialisers should be deserialising this, not Hyperion.  I also thought that some of my own Entities seem to occasionally fall into Hyperion, but I realised that I'm using Protobuf for Snapshot, but Hyperion for some Journaled messages (so I need to resolve that)

Here's the complete HOCON [HOCON.txt](https://github.com/akkadotnet/akka.net/files/6753137/HOCON.txt)

**Comment by Arkatufus** [comment]
@andyfurnival Akka.NET might serialize or deserialize the message into protobuf, but the message itself is serialized into binary using the serializer defined inside your HOCON settings. In effect, the protobuff message is an envelope or wrapper around a serialized binary data returned by the serializer for said message type.

So the message itself is still serialized and deserialized using Hyperion, unless you've overriden it with your own protobuff serializer.

**Comment by Aaronontheweb** [comment]
We have a reproduction for this now https://github.com/akkadotnet/akka.net/pull/5138#issuecomment-878305169

**Comment by Aaronontheweb** [comment]
Should be resolved when https://github.com/akkadotnet/akka.net/pull/5140 is merged

**Comment by Aaronontheweb** [comment]
Ah, this was fixed part of v1.4.22.
